plugins/aws/providers.go

```diff
@@ -0,0 +1,142 @@
+	stsCacheWriter
+}
+
+func (p assumeRoleProvider) Retrieve(ctx context.Context) (aws.Credentials, error) {
+	credentials, err := p.AssumeRoleProvider.Retrieve(ctx)
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	err = p.stsCacheWriter.persist(credentials)
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	return credentials, nil
+}
+
+func NewAssumeRoleProvider(awsConfig *confighelpers.Config, in sdk.ProvisionInput, out *sdk.ProvisionOutput) aws.CredentialsProvider {
+	roleCacheKey := getRoleCacheKey(awsConfig.RoleARN)
+	if in.Cache.Has(roleCacheKey) {
+		return NewStsCacheProvider(roleCacheKey, in.Cache)
+	}
+
+	if awsConfig.HasMfaSerial() && awsConfig.MfaToken != "" {
+		return &assumeRoleProvider{
+			AssumeRoleProvider: confighelpers.AssumeRoleProvider{
+				StsClient:         getSTSClient(awsConfig.Region, NewMFASessionTokenProvider(awsConfig, in, out)),
+				RoleARN:           awsConfig.RoleARN,
+				RoleSessionName:   awsConfig.RoleSessionName,
+				ExternalID:        awsConfig.ExternalID,
+				Duration:          900 * time.Second, // minimum duration of 15 minutes
+				Tags:              awsConfig.SessionTags,
+				TransitiveTagKeys: awsConfig.TransitiveSessionTags,
+				SourceIdentity:    awsConfig.SourceIdentity,
+				Mfa:               &confighelpers.Mfa{},
+			},
+			stsCacheWriter: NewStsCacheWriter(roleCacheKey, out.Cache),
+		}
+	}
+
+	return &assumeRoleProvider{
+		AssumeRoleProvider: confighelpers.AssumeRoleProvider{
+			StsClient:         getSTSClient(awsConfig.Region, NewMasterCredentialsProvider(in.ItemFields)),
+			RoleARN:           awsConfig.RoleARN,
+			RoleSessionName:   awsConfig.RoleSessionName,
+			ExternalID:        awsConfig.ExternalID,
+			Duration:          900 * time.Second, // minimum duration of 15 minutes
+			Tags:              awsConfig.SessionTags,
+			TransitiveTagKeys: awsConfig.TransitiveSessionTags,
+			SourceIdentity:    awsConfig.SourceIdentity,
+			Mfa:               &confighelpers.Mfa{},
+		},
+		stsCacheWriter: NewStsCacheWriter(roleCacheKey, out.Cache),
+	}
+}
+
+type mfaSessionTokenProvider struct {
+	confighelpers.SessionTokenProvider
+	stsCacheWriter
+}
+
+func (p mfaSessionTokenProvider) Retrieve(ctx context.Context) (aws.Credentials, error) {
+	credentials, err := p.SessionTokenProvider.Retrieve(ctx)
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	err = p.stsCacheWriter.persist(credentials)
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	return credentials, nil
+}
+
+func NewMFASessionTokenProvider(awsConfig *confighelpers.Config, in sdk.ProvisionInput, out *sdk.ProvisionOutput) aws.CredentialsProvider {
+	if in.Cache.Has(mfaCacheKey) {
+		return NewStsCacheProvider(mfaCacheKey, in.Cache)
+	}
+
+	return &mfaSessionTokenProvider{
+		SessionTokenProvider: confighelpers.SessionTokenProvider{
+			StsClient: getSTSClient(awsConfig.Region, NewMasterCredentialsProvider(in.ItemFields)),
+			Duration:  900 * time.Second, // minimum duration of 15 minutes,
```

Comment:
I've seen this pop up in 3 places now. Let's make sure to use one source of truth for this.

Reply:
Done, thanks!